### PR TITLE
Replace `toml_map_to_field` and `toml_remap` with traits to map between `InputValue`s and `TomlTypes`

### PR DESCRIPTION
**Pull request by TomAFrench**
# Related issue(s)

# Description

## Summary of changes

I've replaced the functions `toml_map_to_field` and `toml_remap` with implementations of `TomlTypes::from<InputValue>()` and `InputValue::try_from<TomlTypes>()`, i.e. rather than mapping the entire set of arguments at once we specify how to map a single input from one type to another.

We can then just iterate over each argument and map them individually.

## Dependency additions / changes

N/A

## Test additions / changes

N/A

# Checklist

- [x] I have tested the changes locally.
- [x] I have formatted the changes with [Prettier](https://prettier.io/) and/or `cargo fmt` with default settings.
- [x] I have [linked](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue) this PR to the issue(s) that it resolves.
- [x] I have reviewed the changes on GitHub, line by line.
- [x] I have ensured all changes are covered in the description.

# Additional context

### CODE REVIEW

File: crates/noirc_abi/src/input_parser/toml.rs [hunk 1]
```diff
@@ -82,67 +113,27 @@ fn toml_map_to_field(
...
-                let array_elements: Vec<_> = arr_str
-                    .into_iter()
-                    .map(|elem_str| parse_str(&elem_str).unwrap().unwrap())
-                    .collect();
+                let array_elements = try_vecmap(arr_str, |elem_str| match parse_str(&elem_str) {
+                    Ok(Some(elem_value)) => Ok(elem_value),
+                    Ok(None) => {
+                        Err(InputParserError::ParseStr("undefined array element".to_string()))
+                    }
+                    Err(err) => Err(err),
```
Comment (jfecher) [inline crates/noirc_abi/src/input_parser/toml.rs:121]: Nitpicking but we can replace this match with

```rs
parse_str(&elem_str)?.ok_or_else(|| 
    InputParserError::ParseStr("undefined array element".to_string()
)
```

**Review by jfecher — APPROVED** [review]
Looks great, just one style comment that you can choose to ignore if you wish

File: crates/noirc_abi/src/input_parser/toml.rs (hunk 1 above)
Comment (TomAFrench) [reply crates/noirc_abi/src/input_parser/toml.rs:121]: Come to think of it, it'd likely be better to have `parse_str` return a proper error type for empty strings. We currently need to have special handling for them anyway (L91-95) and when string support is added we won't need to parse the string anyway.

**Comment by TomAFrench** [comment]
Looks like this will end up incompatible with https://github.com/noir-lang/noir/pull/622 due to it requiring context about whether a particular value is a field or a string which we can't pass to `try_from`. We could create some ad-hoc conversion method however where we can pass in the desired `AbiType`.

**Comment by jfecher** [comment]
I believe the reason PR #622 adds the type-aware conversions is to keep allowing `"0xabcd"` in the toml to decode to an integer or field. Does toml not allow unquoted hexadecimal literals like `0xabcd` or is there some maximum bits limitation in toml I'm not aware of? If toml allows the unquoted hex literals my vote would be to remove support for the quoted ones.

**Comment by vezenovm** [comment]
> I believe the reason PR #622 adds the type-aware conversions is to keep allowing `"0xabcd"` in the toml to decode to an integer or field. Does toml not allow unquoted hexadecimal literals like `0xabcd` or is there some maximum bits limitation in toml I'm not aware of? If toml allows the unquoted hex literals my vote would be to remove support for the quoted ones.

To my knowledge toml does not allow unquoted hexadecimal literals and that is why we use strings for hexadecimal inputs. If toml does allow unquoted hex literals I agree we should just remove support for the quoted ones to avoid confusion.

**Comment by TomAFrench** [comment]
> Does toml not allow unquoted hexadecimal literals like 0xabcd or is there some maximum bits limitation in toml I'm not aware of? If toml allows the unquoted hex literals my vote would be to remove support for the quoted ones.

It does support hex representations of integers however integers are limited to values which fit in an `i64` so without accepting string representations we wouldn't be able to specify a large range of fields.

**Comment by TomAFrench** [comment]
I've rebased this back on top of master to support string types.

**Review by jfecher — DISMISSED** [review]
👍

**Review by vezenovm — DISMISSED** [review]

**Review by jfecher — APPROVED** [review]

**Pull request merged**